plugin/world_plugin.cpp

```diff
@@ -0,0 +1,111 @@
+
+#include "message/input/gazebo/Command.pb.h"
+#include "message/input/gazebo/Simulation.pb.h"
+#include "nuclear_network.h"
+
+namespace gazebo {
+class NUbotsWorldPlugin : public WorldPlugin {
+public:
+    // Inherit constructors
+    using WorldPlugin::WorldPlugin;
+
+    template <typename T>
+    using Network = NUClear::dsl::word::Network<T>;
+
+    // Make sure to unbind the reaction handle on destruction
+    virtual ~NUbotsWorldPlugin() {
+        handle.unbind();
+    }
+
+    /**
+     * The load function is called by Gazebo when the plugin is inserted into simulation
+     * @param _model A pointer to the model that this plugin is attached to
+     * @param _sdf   A pointer to the plugin's SDF element
+     */
+    void Load(physics::WorldPtr _world, sdf::ElementPtr _sdf) {
+
+        // Last update is when we load
+        last_update = std::chrono::steady_clock::now() - std::chrono::milliseconds(10);
+
+        // Just output a message for now
+        gzdbg << "Attaching a NUbots world plugin to [" << _world->Name() << "]" << std::endl;
+
+        // Configure the NUClear network
+        auto net_config              = std::make_unique<NUClear::message::NetworkConfiguration>();
+        net_config->name             = _sdf->Get<std::string>("nuclearnet_name", "gazebo").first;
+        net_config->announce_address = _sdf->Get<std::string>("nuclearnet_address", "239.226.152.162").first;
+        net_config->announce_port    = _sdf->Get<int>("nuclearnet_port", 7447).first;
+        get_reactor().emit<NUClear::dsl::word::emit::Direct>(net_config);
+
+        // Store the model pointer for convenience
+        this->world = _world;
+
+        // When we get an update event, publish the ball position and velocity
+        update_connection = event::Events::ConnectWorldUpdateBegin(std::bind(&NUbotsWorldPlugin::update_world, this));
+
+        handle = get_reactor().on<Network<message::input::gazebo::Command>>().then(
+            [this](const message::input::gazebo::Command& c) {
+                // Store this command ready for the next update as we don't want to mess with gazebos threading
+                std::lock_guard<std::mutex> lock(command_mutex);
+                command_queue.push_back(c);
+            });
+    }
+
+
+private:
+    void update_world() {
+        // Run all our commands in the command queue and then clear it
+        {
+            std::lock_guard<std::mutex> lock(command_mutex);
+            for (const auto& c : command_queue) {
+                command(c);
+            }
+            command_queue.resize(0);
+        }
+
+        // Only send packets at 100Hz
+        auto now = std::chrono::steady_clock::now();
+        if (now - last_update > std::chrono::milliseconds(10)) {
+            last_update += std::chrono::milliseconds(10);
+            // Make our message
+            auto msg = std::make_unique<message::input::gazebo::Simulation>();
+            msg->set_sim_time(this->world->SimTime().Double());
```

Comment:
Something that would be good for the other systems is to send packets at a rate based on simtime rather than realtime. That way we send a new sensors packet 90/s using sim time instead.

Reply:
Done!